### ISSUE TITLE
The optimizer is biased towards reviews with low stabilities

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#261]
For almost any collection, much greater number of reviews have low stability as compared to a high stability. For example, in the following image, consider the number of predictions having stability less than and greater than 63 days respectively.

<img width=512px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/198ed256-fee5-40f7-bd40-3739ddc30a8b">

This means that the amount of data available for high stability reviews is much lesser than the amount of data available for low stability reviews.

Since the optimizer tries to minimize the total loss, it means that the optimizer would try to produce such values of the parameters that minimize the loss for low stability reviews even if they cause large loss for the high stability reviews. This is probably the reason that the model doesn't work well where the stability is very high.

It is possible to do something (such as adding weights to the reviews) so that the performance is equally good for low stability and high stability reviews.

I don't know how much of a work it is to push the same change to multiple branches at once. But, at least push this to the branches with the new forgetting curve [(t/S)^a] and the sm-stabilization curve.

If the change appears to produce good results in the experimental versions, I recommend you to also push the change to the stable version of the optimizer even if we haven't got good results in trying to improve the algorithm yet.

By the way, the log loss and the RMSE don't indicate this kind of bias. So, is there any other metric that we can use to expose such bias? Otherwise, the stability calibration curve is also a good way to show this bias. The only thing is that it is much easier to compare values rather than comparing two calibration graphs.

**Comment by user1823** [comment]
I don't know what you understood from my comment. But I wanted you to improve the training process so that the performance of the model is equally good for low stability and high stability reviews.

**Comment by L-M-Sherlock** [comment]
I get it. But we need to trade off. The number of reviews with low stability is greater than those with high stability, because we review more in low stability. So the accurate prediction of reviews with low stability is more important because we frequently review them. Is it wrong?

**Comment by Expertium** [comment]
@user1823 I don't think that the number of reviews with "low" stability should be equal to the number of reviews with "high" stability (also, your cutoff point of 63 days seems arbitrary, the cutoff depends on the material of your collection and how long ago you have started learning). The more you use any spaced repetition algorithm, the more stability will grow. The greater the value of S, the more time (and reps) it takes to reach that value. Unless you have been using spaced repetition for 10+ years and ran out of new cards to learn, you should expect to see very few cards with very high S.
Also, IMO, this calibration graph isn't nearly as useful as the old one.

**Comment by user1823** [comment]
> I get it. But we need to trade off. The number of reviews with low stability is greater than those with high stability, because we review more in low stability. So the accurate prediction of reviews with low stability is more important because we frequently review them. Is it wrong?

I got your point.

But, I think that an ideal model should perform equally well for low stability and high stability reviews.

Also, if the model performs poorly for high stability reviews, it would result in more lapses and more high stability reviews would return as low stability ones. This would increase the workload greatly.

Also, if the model performs poorly for high stability reviews, it means that the long-term retention is hampered. This is not good because good long-term retention is the ultimate goal of using spaced repetition.

By the way, whatever the case may be, we should at least try doing it. Maybe, we could be surprised by the results. (Maybe the results for the low stability cards are not affected that much but the results for the high stability cards are greatly improved).

**Comment by L-M-Sherlock** [comment]
> But, I think that an ideal model should perform equally well for low stability and high stability reviews.

Ideally, it should be. But there is another problem. For those cards with high stability, we have more stochastic influence happen during the long days before we review them in the next time. So the retention of reviews with high stability is inaccurate/unstable/unclear by itself.

**Comment by user1823** [comment]
> But there is another problem. For those cards with high stability, we have more stochastic influence happen during the long days before we review them in the next time. So the retention of reviews with high stability is inaccurate/unstable/unclear by itself.

That's also a good point.

But, I still think that there is nothing wrong in trying. I have a strong gut feeling that trying to normalise the performance of the model for low stability and high stability reviews would produce much better results than we are currently getting.

Just for example, consider the following:

For low stabilities (say, less than 20 days), the value of the term S^w7 in recall stability would not be much affected if the w7 changes from say, -0.12 to -0.17.

But, for large stabilities (say, more than 150 days), the value of the term S^w7 in recall stability would be much more affected if the w7 changes from say, -0.12 to -0.17.

**Comment by Expertium** [comment]
> I have a strong gut feeling that trying to normalise the performance of the model for low stability and high stability reviews would produce much better results than we are currently getting.

That's interesting, do you have a specific formula in mind?

**Comment by user1823** [comment]
> That's interesting, do you have a specific formula in mind?

The only thing I can think of currently is the following: 
- First, we divide the reviews into several bins based on their stabilities. For example, the first bin contains reviews with stabilities between 1-5 days, the second bin contains reviews with stabilities between 6-10 days and so on.
- Then, the new loss can be defined as the sum of the current loss for each bin divided by the number of reviews in that bin.
Then, we will make the optimizer minimize this new loss instead of the current loss.

This way, the high stability and the low stability reviews would have a similar contribution in the training process.

I am also open to other suggestions to normalise the performance of the model for high stability and low stability reviews.

**Comment by L-M-Sherlock** [comment]
> Good news - I finally found a way to get the benefits of exp(initial S) without the downsides.
> Before I tried this:
> `new_s = torch.exp(self.w[0] + self.w[1] * (X_rating - self.w[14]))`
> It drastically improved RMSE for a few easy decks, made it worse for a few hard decks, and had either a small positive or a small negative impact on most decks, so overall it wasn't great. I added another parameter, like this:
> `new_s = self.w[18] * torch.exp(self.w[0] + self.w[1] * (X_rating - self.w[14]))`
> Clampings look like this:
> ```
> w[0] = w[0].clamp(-4, 4)
> w[1] = w[1].clamp(-4, 4)
> w[18] = w[18].clamp(0.01, 10)
> ```
> This new formula for calculating initial S improved RMSE, and I recommend implementing it in 4.0 Alpha.
> ![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/ba78a816-0ed8-461d-86a1-993e208b35e3)
> 
> Strangely, it made results slightly worse for Sherlock's collection. This is strange because if you replace `self.w[18] * torch.exp(self.w[0] + self.w[1] * (X_rating - self.w[14]))` with just `torch.exp(self.w[0] + self.w[1] * (X_rating - self.w[14]))`, which is equivalent to setting self.w[18] to 1, RMSE gets better. In other words, it's possible to achieve lower loss with this new formula, but the optimizer can't find the right parameters for some reason.
> I'm surprised how much changing the initial S affects results, especially for very easy decks. I guess it's because of the "ripple effect" - since future estimates of S depend on the initial estimate, an inaccuracy in the initial estimate will disturb all future estimates.
> 
> _Originally posted by @Expertium in https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1597811370_

I think if the model doesn't predict the initial stability (usually low stability) precisely, it would also have poor performance on the high stability.

**Comment by Expertium** [comment]
> I think if the model doesn't predict the initial stability (usually low stability) precisely, it would also have poor performance on the high stability. So this issue could be not a right problem.

I don't understand what you're trying to say, can you elaborate?

**Comment by L-M-Sherlock** [comment]
> I don't understand what you're trying to say, can you elaborate?

I just quoted your results. I mean this issue could be closed.

**Comment by Expertium** [comment]
Ah, my bad. For some reason I thought that you thought that this new formula is bad.

**Comment by user1823** [comment]
It is convincing enough. So, I am closing the issue.
